macros/src/main/scala/scala/tools/abide/traversal/WarningsMacro.scala

```diff
@@ -0,0 +1,73 @@
+package scala.tools.abide.traversal
+
+import scala.reflect.macros._
+import scala.language.experimental.macros
+
+/**
+ * WarningExtractor
+ *
+ * Provide the extraction macro that uses the placeholder [[WarningExtractor.tree]] value in the
+ * [[WarningExtractor]] class with to create a function from Tree to String by replacing the
+ * dummy tree with the function parameter.
+ *
+ * To get the String we use the StringContext that we assume is the macro caller and
+ * apply the `StringContext.s` method to the transformed arguments. We also assume all
+ * references to `_.this.tree` refer to the [[WarningExtractor.tree]] value as checking for
+ * enclosing context is non-trivial and discouraged.
+ */
+object WarningExtractor {
```

Comment:
Well, the interpolator actually returns a function of type `Tree => String` so calling it inside a warning creation wouldn't make much sense. Might as well use the `s` interpolator in that case. In fact, you can define the `warning` member as a function directly if you want to use the `nok(tree)` syntax.

If you use the `w` interpolator somewhere where `this.tree` is masked, then indeed, you won't get dynamic replacement, but the interpolator doesn't make much sense anywhere else than top-level so I don't really see this as an issue. The only corner case I can imagine is if people start defining inner-classes with tree members and `w` interpolators inside these where the `tree` binder could be masked, but this doesn't sound like normal rule usage and should probably be discouraged anyway.